File: experimento-1-acl-kyc/acl-worker/src/infra/cola_reconciliacion.py
"""
Cola de reconciliación diferida del Consolidador KYC (ver "Extensión de
diseño: Consolidador KYC (reconciliación diferida)" en
../../../DISENO-EXPERIMENTOS.md). El ACL Worker es el PRODUCTOR: cada vez
que ServicioVerificacion resuelve una llamada como "degradado", encola un
job aquí para que el Consolidador KYC reintente más tarde contra el ACL
Worker (nunca contra el proveedor KYC directo).

Migrado de Node.js/BullMQ a Python/RQ (Redis Queue) — ver "Decisión sobre la
librería de cola" en el README de esta carpeta para la justificación
completa. Nota importante de RQ: el PRODUCTOR encola la función por su
*import path como string* ("jobs.reconciliar_cliente"); el código real de
esa función vive únicamente en `consolidador-kyc/src/jobs.py` — el ACL
Worker nunca la importa ni necesita conocerla más allá del nombre.

Requisito no negociable del diseño: este encolado es fire-and-forget y NUNCA
puede tumbar ni retrasar la respuesta que el ACL Worker ya le dio a UNDER.
Si Redis está caído, el fallo se loguea y se ignora — no se propaga ninguna
excepción hacia ServicioVerificacion.
"""

import logging
import time

from redis import Redis
from rq import Queue, Retry

from ..config import config

logger = logging.getLogger("acl-worker.cola-reconciliacion")

NOMBRE_COLA_RECONCILIACION = "kyc-reconciliacion"

_conexion: Redis | None = None
_cola: Queue | None = None
_fallo_inicializacion = False


def _obtener_cola() -> Queue | None:
    global _conexion, _cola, _fallo_inicializacion
    if _cola is not None:
        return _cola
    if _fallo_inicializacion:
        return None  # ya se intentó y falló; no reintentar en cada llamada

    try:
        _conexion = Redis.from_url(
            config.redis_url,
            socket_connect_timeout=2,
            socket_timeout=2,
        )
        _cola = Queue(NOMBRE_COLA_RECONCILIACION, connection=_conexion)
        return _cola
    except Exception as error:  # noqa: BLE001 - red de seguridad no bloqueante
        _fallo_inicializacion = True
        logger.error("no se pudo inicializar la cola (no bloqueante): %s", error)
        return None


def encolar_reconciliacion(cliente_id: str) -> None:
    """
    Encola `(clienteId, timestamp)` en `kyc-reconciliacion`. Se llama desde
    un hilo aparte (`asyncio.to_thread`, ver ServicioVerificacion) para no
    bloquear el event loop — es fire-and-forget puro, igual que el `.then()`
    sin `await` de la versión Node/BullMQ.

    Política de reintentos: 5 intentos totales (1 inicial + 4 reintentos),
    backoff exponencial 2s/4s/8s/16s — mismos números que la versión
    Node/BullMQ (`attempts: 5, backoff: { type: 'exponential', delay: 2000 }`).
    Si se agotan, RQ mueve el job a su `FailedJobRegistry` automáticamente
    (equivalente a la "DLQ" del diseño) — no se construye nada adicional.
    """
    try:
        cola = _obtener_cola()
        if cola is None:
            return

        cola.enqueue(
            "jobs.reconciliar_cliente",
            cliente_id,
            time.time() * 1000,
            retry=Retry(max=4, interval=[2, 4, 8, 16]),
            result_ttl=0,
        )
        logger.info("job encolado para clienteId=%s", cliente_id)
    except Exception as error:  # noqa: BLE001 - nunca debe llegar a ServicioVerificacion
        logger.error("fallo al encolar clienteId=%s (no bloqueante): %s", cliente_id, error)

File: experimento-1-acl-kyc/acl-worker/src/infra/ColaReconciliacion.ts
import { Queue } from 'bullmq';
import IORedis from 'ioredis';
import { config } from '../config';

/**
 * Cola de reconciliación diferida del Consolidador KYC (ver "Extensión de
 * diseño: Consolidador KYC (reconciliación diferida)" en
 * ../../../DISENO-EXPERIMENTOS.md). El ACL Worker es el PRODUCTOR: cada vez
 * que ServicioVerificacion resuelve una llamada como "degradado", encola un
 * job aquí para que el Consolidador KYC reintente más tarde contra el ACL
 * Worker (nunca contra el proveedor KYC directo).
 *
 * Requisito no negociable del diseño: este encolado es fire-and-forget y
 * NUNCA puede tumbar ni retrasar la respuesta que el ACL Worker ya le dio a
 * UNDER. Si Redis está caído, el fallo se loguea y se ignora — no se lanza
 * ninguna excepción hacia ServicioVerificacion.
 */

export const NOMBRE_COLA_RECONCILIACION = 'kyc-reconciliacion';

let colaSingleton: Queue | null = null;
let colaFalloInicializacion = false;

function obtenerCola(): Queue | null {
  if (colaSingleton) return colaSingleton;
  if (colaFalloInicializacion) return null; // ya se intentó y falló; no reintentar en cada llamada

  try {
    const conexion = new IORedis(config.redisUrl, {
      // Requerido por BullMQ para conexiones usadas por sus componentes.
      maxRetriesPerRequest: null,
      enableReadyCheck: false,
      lazyConnect: false,
      // Reintenta la conexión de fondo, pero nunca lanza hacia arriba: el
      // listener 'error' de abajo es lo único que se ejecuta ante un fallo.
      retryStrategy: (intento) => Math.min(intento * 500, 5000),
    });

    conexion.on('error', (error) => {
      console.error(
        `[acl-worker][cola-reconciliacion] error de conexión a Redis (no bloqueante, no afecta respuestas a UNDER): ${error.message}`,
      );
    });

    colaSingleton = new Queue(NOMBRE_COLA_RECONCILIACION, { connection: conexion });
    return colaSingleton;
  } catch (error) {
    colaFalloInicializacion = true;
    console.error(
      `[acl-worker][cola-reconciliacion] no se pudo inicializar la cola (no bloqueante): ${(error as Error).message}`,
    );
    return null;
  }
}

/**
 * Encola `{ clienteId, timestamp }` en `kyc-reconciliacion`. Se llama SIN
 * `await` desde ServicioVerificacion — es fire-and-forget puro. La política
 * de reintentos (5 intentos, backoff exponencial) se define aquí, en el
 * productor, porque son opciones del job (`queue.add`), no del worker que
 * lo consume (el Consolidador KYC). Si se agotan, BullMQ mueve el job a su
 * *failed set* automáticamente (la "DLQ" del diseño) — no se construye nada
 * adicional para eso.
 */
export function encolarReconciliacion(clienteId: string): void {
  try {
    const cola = obtenerCola();
    if (!cola) return;

    cola
      .add(
        'reconciliar',
        { clienteId, timestamp: Date.now() },
        {
          attempts: 5,
          backoff: { type: 'exponential', delay: 2000 },
          removeOnComplete: true,
          removeOnFail: false, // deja el job visible en el failed set (DLQ) tras agotar intentos
        },
      )
      .then(() => {
        console.log(`[acl-worker][cola-reconciliacion] job encolado para clienteId=${clienteId}`);
      })
      .catch((error) => {
        console.error(
          `[acl-worker][cola-reconciliacion] fallo al encolar clienteId=${clienteId} (no bloqueante): ${(error as Error).message}`,
        );
      });
  } catch (error) {
    // Red de seguridad final: ningún fallo de esta función debe propagarse
    // hacia ServicioVerificacion/el fallback del Circuit Breaker.
    console.error(
      `[acl-worker][cola-reconciliacion] fallo inesperado al encolar clienteId=${clienteId} (no bloqueante): ${(error as Error).message}`,
    );
  }
}
